scio-extra/src/main/scala/com/spotify/scio/extra/Sparkey.scala

```diff
@@ -0,0 +1,173 @@
+/*
+ * Copyright 2017 Spotify AB.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.spotify.scio.extra
+
+import java.io._
+import java.net.URI
+import java.nio.ByteBuffer
+import java.nio.file.{Paths, Files => JFiles}
+import java.util.UUID
+
+import com.google.common.base.Charsets
+import com.google.common.hash.Hashing
+import com.spotify.scio.ScioContext
+import com.spotify.scio.util.{CallSites, ScioUtil}
+import com.spotify.scio.values.{SCollection, SideInput}
+import com.spotify.sparkey.{CompressionType, SparkeyReader, Sparkey => JSparkey}
+import org.apache.beam.sdk.options.PipelineOptionsFactory
+import org.apache.beam.sdk.transforms.{DoFn, View}
+import org.apache.beam.sdk.util.GcsUtil.GcsUtilFactory
+import org.apache.beam.sdk.util.gcsfs.GcsPath
+import org.apache.beam.sdk.values.PCollectionView
+
+import scala.collection.JavaConverters._
+
+object Sparkey {
+
+  /**
+   * Represents the base URI for a Sparkey index and log file, either on the local file
+   * system or on GCS. For GCS, the URI should be in the form
+   * 'gs://<bucket>/<path>/<sparkey-prefix>'. For local files, the URI should be in the form
+   * '/<path>/<sparkey-prefix>'. Note that the URI must not be a folder or GCS bucket as the URI is
+   * a base path representing two files - <sparkey-prefix>.spi and <sparkey-prefix>.spl.
+   */
+  trait SparkeyUri {
+    val basePath: String
+    def getReader(): SparkeyReader
+    override def toString(): String = basePath
+  }
+
+  object SparkeyUri {
+    def apply(path: String): SparkeyUri =
+      if (ScioUtil.isGcsUri(new URI(path))) new GcsSparkeyUri(path) else new LocalSparkeyUri(path)
+  }
+
+  implicit class SparkeyScioContext(val self: ScioContext) extends AnyVal {
```

Comment:
It's a scala 2.11+ thing, `extends AnyVal` make this a value class and optimizes away the `new SparkeyScioContext` constructor step.